src/main/groovy/lang/ObjectRange.java

```diff
@@ -488,37 +500,61 @@ public void remove() {
 
     @Override
     public List step(int step) {
+        return step((Number) step);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List step(Number step) {
         final IteratorClosureAdapter adapter = new IteratorClosureAdapter(this);
         step(step, adapter);
         return adapter.asList();
     }
 
     /**
-     * Increments by one
+     * Increments by step size
      *
      * @param value the value to increment
-     * @return the incremented value
+     * @param step  the value to increment by
+     * @return the incremented value or null, if there isn't any
      */
-    private static Object increment(Object value) {
+    protected static Comparable increment(Comparable value, Number step) {
         if (value instanceof Number) {
-            return plus((Number) value, 1);
-        } else {
-            return InvokerHelper.invokeMethod(value, "next", null);
+            return (Comparable) plus((Number) value, step);
+        }
+
+        for (int i = 0; compareLessThan(i, step); i++) {
```

Comment:
(0.1..0.9) == [0.1]. That's not intuitive, but that's the definition according to javadoc. Call that historic reasons. It might be nice to have Range objects with variable item distance, with a syntax such as 1..2..8, 0..5..100, 0.0..0.1..0.9. 

But creating a range with a variable item distance and traversing through an interval (not range) using a flexible iteration distance are very different things, and it seems to me this PR mixes them up.

Also see this for similar issues in Swift:
* http://oleb.net/blog/2015/09/swift-ranges-and-intervals/
* https://www.hackingwithswift.com/example-code/language/using-stride-to-loop-over-a-range-of-numbers